core/primitives/src/stateless_validation.rs

```diff
@@ -17,10 +18,36 @@ use near_primitives_core::types::{AccountId, Balance, BlockHeight, ShardId};
 /// This is a messy workaround until we know what to do with NEP 483.
 type SignatureDifferentiator = String;
 
-/// Signable
+/// Represents bytes of encoded ChunkStateWitness.
+/// For now encoding is raw borsh serialization, later we plan
+/// adding compression on top of that.
 #[derive(Debug, Clone, PartialEq, Eq, BorshSerialize, BorshDeserialize)]
-pub struct ChunkStateWitness {
-    pub inner: ChunkStateWitnessInner,
+pub struct EncodedChunkStateWitness(Box<[u8]>);
+
+impl EncodedChunkStateWitness {
+    pub fn encode(witness: &ChunkStateWitness) -> Self {
+        Self(borsh::to_vec(witness).expect("borsh serialization should not fail").into())
+    }
+
+    pub fn decode(&self) -> Result<ChunkStateWitness, std::io::Error> {
+        ChunkStateWitness::try_from_slice(&self.0)
+    }
+
+    pub fn size_bytes(&self) -> usize {
+        self.0.len()
+    }
+
+    pub fn as_slice(&self) -> &[u8] {
+        &self.0
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, BorshSerialize, BorshDeserialize)]
+pub struct SignedChunkStateWitness {
```

Comment:
Could we create a `SignedChunkStateWitness::new(witness: &ChunkStateWitness, signer: &dyn ValidatorSigner) -> Self`?

Reply:
Maybe add a `fn witness_size()` here as well

Reply:
added `new`: 680d7effcbb9ef17db780119f64f54645acfafa9
regarding `fn witness_size()`: we can use `signed_witness.witness_bytes.size_bytes()`, it is not used in many places, so let's keep it like this for now